stumpy/scrump.py

```diff
@@ -103,67 +109,113 @@ def _compute_PI(
     for i in indices[start:stop]:
         Q = T_A[i : i + m]
         QT[:] = core._sliding_dot_product(Q, T_B)
-        # Update P[i] relative to all T[j : j + m]
         squared_distance_profile[:] = core._mass(Q, T_B, QT, μ_Q[i], σ_Q[i], M_T, Σ_T)
         squared_distance_profile[:] = np.square(squared_distance_profile)
         if excl_zone is not None:
             zone_start = max(0, i - excl_zone)
             zone_stop = min(l, i + excl_zone)
             squared_distance_profile[zone_start : zone_stop + 1] = np.inf
 
-            # only for self-join
-            mask = squared_distance_profile < P_squared[thread_idx]
-            P_squared[thread_idx][mask] = squared_distance_profile[mask]
-            I[thread_idx][mask] = i
+        if excl_zone is not None:  # self-join
+            # note: S_index = T[index: index + m]
+            # `v = squared_distance_profile[idx]` is (the square of)
+            # `dist(S_i, S_idx)`, which is the same as `dist(S_idx, S_i)`. So,
+            # `squared_distance_profile[idx]` is (the square of) distane from `S_idx`
+            # to one of its neighbors, `S_i`. Therefore, the value `v` can be
+            # used to update the TopK of `S_idx`, stored "ascendingly" in
+            # `P_squared[thread_idx, idx, :]`.
+
+            # `P_squared[thread_idx, idx, :]` in inf for those `idx` that are in
+            # the trivial zone, including the `i` itself. Those are not updated here.
+            IDX = np.flatnonzero(
+                squared_distance_profile < P_squared[thread_idx, :, -1]
+            )
+            for idx in IDX:
+                d_squared = squared_distance_profile[idx]
+                pos = np.searchsorted(
+                    P_squared[thread_idx, idx], d_squared, side="right"
+                )
+                core._shift_insert_at_index(P_squared[thread_idx, idx], pos, d_squared)
+                core._shift_insert_at_index(I[thread_idx, idx], pos, i)
 
-        I[thread_idx, i] = np.argmin(squared_distance_profile)
-        P_squared[thread_idx, i] = squared_distance_profile[I[thread_idx, i]]
-        if P_squared[thread_idx, i] == np.inf:  # pragma: no cover
-            I[thread_idx, i] = -1
+        # find EXACT (not approx.) value of `P_squared[thread_idx, i, 0]`
+        nn_of_i = np.argmin(squared_distance_profile)
```

Comment:
I used the name `nn_of_i` to improve readability and emphasize the fact that this is actually the (first) nearest neighbor of `i`. Also, I avoided `i_nn` as this (may?) confuse the reader that this is going to be the `i`-th neighbor. What do you think? We just use this name in the next few lines.

Reply:
Honestly, I think the confusion is that we handled the top-k for the `j`th subsequence(s) first before we handled the top-k for the `i`th subsequence. In my head, this seems like a non-obvious thing to do. Hence the "order" is bad as it feels like we are flipping back and forth. Instead, where possible, we should always try to handle the obvious `i` case first, then add a comment that we can also do `j`, and then add the code for `j`

Reply:
> Hence the "order" is bad as it feels like we are flipping back and forth. Instead, where possible, we should always try to handle the obvious `i` case first, then add a comment that we can also do top-k for `j`, and then add the code for `j`

I see. Since the outer `for-loop` has the iterator `i`, it makes sense to take care of `i` first. And, then do (top-k) update of other indices.

> However, if you stick to finishing all of your business with `i` and handle its nearest neighbor at the same time then it will be clear that `nn` or even `idx` is referring to the nearest neighbor to `i`

> Okay, if it allows us to reorder the code and handle i cases first then I accept having the redundant if condition above. But something tells me that there's a better design!

I get it now. Cool! I will try to restructure the lines.

> So, the lesson here is organizing things so that related parts always flow together. Otherwise, you are forcing the reader to context switch and keep track of the `i`th subsequence, then the `j`th subsequence, and then you flip back to `i`. This is really hard to follow. At least, it is for me.

It makes sense! Thanks for the lesson :)